File: python/techai_webutils/src/techai_webutils/execution/gate/gate.py
"""Gate composition: Phase, Gate, GateBuilder, GateFactory, run_gate (async port of gate.go).

Composes AnyJobs into ordered parallel/serial phases with optional stop-on-failure, aggregating
per-phase BatchResults and emitting the gate/phase lifecycle to the ExecutionObserver. Per-item
step events come from each job's own fan-out; the gate owns only the gate/phase boundaries.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol, Self, runtime_checkable

from techai_webutils.core.interfaces.execution import BatchResult, NopObserver
from techai_webutils.execution.engine.run import merge_batches, run_job_group

if TYPE_CHECKING:
    from techai_webutils.core.interfaces.execution import AnyJob, ExecutionObserver


@dataclass(frozen=True, slots=True)
class Phase:
    """A named group of jobs run either in parallel or serially (mirrors Go ``gate.Phase``)."""

    name: str
    """Identifies the phase in observer events and result grouping."""

    jobs: tuple[AnyJob, ...]
    """The AnyJobs that make up this phase."""

    parallel: bool
    """Selects concurrent (True) vs sequential-stop-at-first-failure (False) execution."""


@dataclass(frozen=True, slots=True)
class Gate:
    """An ordered sequence of phases with optional stop-on-failure (mirrors Go ``gate.Gate``)."""

    name: str
    """Identifies the gate in observer events and the aggregate outcome."""

    phases: tuple[Phase, ...]
    """The ordered phases executed by run_gate."""

    observer: ExecutionObserver
    """Receives the gate/phase lifecycle events as the gate runs."""

    stop_on_failure: bool = False
    """When True, halts the remaining phases after any phase fails."""


class GateBuilder:
    """Fluent assembler for a Gate (mirrors Go ``gate.GateBuilder``)."""

    def __init__(self, name: str) -> None:
        """Start a builder for a named gate."""
        self._name: str = name
        self._phases: list[Phase] = []
        self._stop_on_failure: bool = False
        self._observer: ExecutionObserver | None = None

    def parallel(self, name: str, *jobs: AnyJob) -> Self:
        """Add a phase whose jobs all run concurrently."""
        self._phases.append(Phase(name=name, jobs=jobs, parallel=True))
        return self

    def serial(self, name: str, *jobs: AnyJob) -> Self:
        """Add a phase whose jobs run one after another, stopping at the first failure."""
        self._phases.append(Phase(name=name, jobs=jobs, parallel=False))
        return self

    def stop_on_failure(self) -> Self:
        """Configure the gate to skip remaining phases after any phase fails."""
        self._stop_on_failure = True
        return self

    def with_observer(self, observer: ExecutionObserver) -> Self:
        """Set the ExecutionObserver for the built gate (defaults to NopObserver)."""
        self._observer = observer
        return self

    def build(self) -> Gate:
        """Return the configured Gate, defaulting the observer to a NopObserver."""
        observer: ExecutionObserver = self._observer if self._observer is not None else NopObserver()
        return Gate(
            name=self._name,
            phases=tuple(self._phases),
            observer=observer,
            stop_on_failure=self._stop_on_failure,
        )


def new_gate(name: str) -> GateBuilder:
    """Return a GateBuilder for a named gate (mirrors Go ``gate.NewGate``)."""
    return GateBuilder(name)


@runtime_checkable
class GateFactory(Protocol):
    """A typed producer of a Gate (mirrors Go ``gate.GateFactory``).

    A worker type that composes a gate implements it so the gate is a modeled value (its
    parameters as fields) rather than a bare builder call; callers invoke ``gate()``.
    """

    def gate(self) -> Gate:
        """Build and return the composed gate."""
        ...


async def run_gate(gate: Gate) -> BatchResult:
    """Run every phase of ``gate`` in order, aggregating results and firing the lifecycle.

    Fires ``on_gate_start`` -> (per phase ``on_phase_start`` -> ``on_phase_complete``) ->
    ``on_gate_complete`` on the gate's observer, and returns the merged BatchResult over all
    phases. When ``stop_on_failure`` is set, the first phase with failures halts the rest.

    Both the gate and each phase fire their ``*_complete`` event from a ``finally`` (with the
    results gathered so far), so that even a phase whose job raises still balances *both* the phase
    and the gate lifecycle before the exception propagates — mirroring Go ``RunGate``, which always
    reports completion. An observer can therefore pair ``on_phase_start``/``on_phase_complete`` (and
    the gate pair) for per-phase timing or resource cleanup without leaking on a raised job.
    """
    loop = asyncio.get_running_loop()
    start = loop.time()
    observer = gate.observer
    observer.on_gate_start(gate.name, [phase.name for phase in gate.phases])

    phase_results: list[BatchResult] = []
    try:
        for phase in gate.phases:
            phase_start = loop.time()
            observer.on_phase_start(phase.name, [job.meta().name for job in phase.jobs])
            # phase_result stays empty if run_job_group raises; the finally still fires
            # on_phase_complete so the phase lifecycle balances (mirrors the gate's finally below).
            phase_result = BatchResult()
            try:
                phase_result = await run_job_group(phase.jobs, parallel=phase.parallel)
                phase_results.append(phase_result)
            finally:
                observer.on_phase_complete(phase.name, phase_result, loop.time() - phase_start)
            if gate.stop_on_failure and phase_result.has_failures:
                break
    finally:
        result = merge_batches(phase_results)
        observer.on_gate_complete(gate.name, result, loop.time() - start)
    return result

File: go/execution/gate/gate.go
// Package gate provides Phase, Gate, GateBuilder, and RunGate for composing
// AnyJobs into ordered phases with stop-on-failure and error aggregation.
package gate

import (
	"context"
	"fmt"
	"time"

	apperr "github.com/gt-tech-ai/knowledge-engine/go/core/errors"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/core/types"
	"github.com/gt-tech-ai/knowledge-engine/go/execution/engine"
)

// GateFactory is a typed producer of a *Gate. CLI workflow types that compose
// a gate implement it so a workflow is a modeled value (parameters as fields)
// rather than a bare builder function. It is a structural contract: callers use
// the concrete workflow type directly; future gate-producing workflows implement
// this interface to signal intent.
//
// §2.3 exemption — execution-engine primitive: a single Gate() → *Gate producer contract,
// not an id-CRUD data-access surface, so it embeds no foundation generic.
type GateFactory interface {
	// Gate builds and returns the composed *Gate.
	Gate() *Gate
}

// Phase is a named group of AnyJobs executed either in parallel or serially.
type Phase struct {
	// Name identifies the phase in observer events and failure messages.
	Name string

	// Jobs are the AnyJobs that make up this phase.
	Jobs []interfaces.AnyJob

	// parallel selects the execution mode: true runs all jobs concurrently,
	// false runs them one after another (stopping at the first failure).
	parallel bool
}

// Gate is an ordered sequence of Phases with optional stop-on-failure behaviour.
// Field order: obs (interface) then name (string) before phases (slice) to keep
// pointer-scan range at 40 bytes instead of 48.
type Gate struct {
	// obs receives the gate/phase/step lifecycle events as the gate runs.
	obs interfaces.ExecutionObserver

	// name identifies the gate in observer events and the wrapping error.
	name string

	// phases are the ordered phases executed by RunGate.
	phases []Phase

	// stopOnFailure, when true, halts the remaining phases after any phase fails.
	stopOnFailure bool
}

// GateBuilder assembles a Gate via a fluent API.
type GateBuilder struct {
	// obs is the observer applied to the built gate; defaults to NopObserver
	// when never set via WithObserver.
	obs interfaces.ExecutionObserver

	// name is the gate name carried into the built Gate.
	name string

	// phases accumulate as Parallel/Serial are called, preserving call order.
	phases []Phase

	// stopOnFailure records whether StopOnFailure was requested.
	stopOnFailure bool
}

// NewGate returns a GateBuilder for a named gate.
func NewGate(name string) *GateBuilder { return &GateBuilder{name: name} }

// Parallel adds a phase in which all jobs run concurrently.
func (b *GateBuilder) Parallel(name string, jobs ...interfaces.AnyJob) *GateBuilder {
	b.phases = append(b.phases, Phase{Name: name, Jobs: jobs, parallel: true})
	return b
}

// Serial adds a phase in which jobs run one after another.
func (b *GateBuilder) Serial(name string, jobs ...interfaces.AnyJob) *GateBuilder {
	b.phases = append(b.phases, Phase{Name: name, Jobs: jobs, parallel: false})
	return b
}

// StopOnFailure configures the gate to skip remaining phases after any failure.
func (b *GateBuilder) StopOnFailure() *GateBuilder {
	b.stopOnFailure = true
	return b
}

// WithObserver sets an ExecutionObserver for all phases.
func (b *GateBuilder) WithObserver(obs interfaces.ExecutionObserver) *GateBuilder {
	b.obs = obs
	return b
}

// Build returns the configured Gate.
func (b *GateBuilder) Build() *Gate {
	obs := b.obs
	if obs == nil {
		obs = engine.NopObserver{}
	}
	return &Gate{name: b.name, phases: b.phases, stopOnFailure: b.stopOnFailure, obs: obs}
}

// Phases returns the ordered phases of the gate for inspection (e.g. in tests).
func (g *Gate) Phases() []Phase { return g.phases }

// RunGateNoRunner runs a gate whose jobs are all runner-less (Runnable-backed via
// job.Adapt). It supplies a no-op CommandRunner so no phase ever receives a nil
// runner; the runner is structurally unused by Runnable-backed jobs. Use it for
// in-process gates (e.g. DB seeding) that compose Runnables rather than commands.
func RunGateNoRunner(ctx context.Context, g *Gate) error {
	return RunGate(ctx, noRunner{}, "", g)
}

// RunGate executes all phases of g in order.
// Each phase's jobs run in parallel or serially per configuration.
// Failures are aggregated into a CodeQualityFailed error (Pattern #21).
// When StopOnFailure is set, the first failed phase halts subsequent phases.
// Returns nil when all phases pass or produce only warnings.
func RunGate(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
	g *Gate,
) error {
	gateStart := time.Now()

	phaseNames := make([]string, len(g.phases))
	for i, p := range g.phases {
		phaseNames[i] = p.Name
	}
	g.obs.OnGateStart(g.name, phaseNames)

	var (
		all     types.StepResults
		gateErr error
	)

	for _, phase := range g.phases {
		phaseResults, err := runPhase(ctx, runner, root, phase, g.obs)
		all = append(all, phaseResults...)

		if err != nil || phaseResults.HasFailures() {
			phaseErr := phaseFailureError(phase.Name, phaseResults)
			gateErr = apperr.Join(gateErr, phaseErr)
			if g.stopOnFailure {
				break
			}
		}
	}

	g.obs.OnGateComplete(g.name, all, time.Since(gateStart))
	if gateErr != nil {
		return apperr.Wrap(gateErr, apperr.CodeQualityFailed, "gate "+g.name+" failed")
	}
	return nil
}

// runPhase executes all jobs in a phase and fires obs.OnPhaseComplete exactly
// once with the phase's aggregate results and measured wall-clock elapsed —
// for both parallel phases (one RunJobGroup call) and serial phases (one call
// per job). It returns the aggregate and the first execution error, if any.
func runPhase(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
	phase Phase,
	obs interfaces.ExecutionObserver,
) (types.StepResults, error) {
	start := time.Now()

	jobNames := make([]string, len(phase.Jobs))
	for i, j := range phase.Jobs {
		jobNames[i] = j.Meta().Name
	}
	obs.OnPhaseStart(phase.Name, jobNames)

	var (
		results  types.StepResults
		phaseErr error
	)
	if phase.parallel {
		results, phaseErr = engine.RunJobGroup(ctx, runner, root, phase.Jobs, obs)
	} else {
		for _, job := range phase.Jobs {
			sr, err := engine.RunJobGroup(
				ctx,
				runner,
				root,
				[]interfaces.AnyJob{job},
				obs,
			)
			results = append(results, sr...)
			if err != nil {
				phaseErr = err
				break
			}
			if sr.HasFailures() {
				break
			}
		}
	}

	obs.OnPhaseComplete(phase.Name, results, time.Since(start))
	return results, phaseErr
}

// phaseFailureError constructs a CodeQualityFailed error aggregating all
// failed step errors in results (Pattern #21).
func phaseFailureError(phaseName string, results types.StepResults) error {
	var errs []error
	for _, r := range results {
		if r.Status == types.StatusFail && r.Error != "" {
			errs = append(
				errs,
				apperr.New(
					apperr.CodeQualityFailed,
					fmt.Sprintf("%s: %s", r.Name, r.Error),
				),
			)
		}
	}
	if len(errs) == 0 {
		return apperr.New(apperr.CodeQualityFailed, "phase "+phaseName+" failed")
	}
	return apperr.Wrap(
		apperr.Join(errs...),
		apperr.CodeQualityFailed,
		"phase "+phaseName+" failed",
	)
}
